metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,541 @@
+              otherQual
+            )
+              +: Enumerator.Guard.apply(
+                if (isFilter) nextCondition
+                else
+                  Term.ApplyUnary.apply(Term.Name.apply("!"), nextCondition)
+              ) +: elems,
+            maybeYieldTerm,
+            newMetalsNames
+          )
+      }
+    }
+    result.getOrElse(List.empty, currentYieldTerm, generatedByMetalsNames)
+  }
+
+  private def extractChainedForYield(
+      perhapseLastName: Option[String],
+      currentYieldTerm: Option[Term],
+      existingForElements: List[Enumerator],
+      termApply: Term.Apply,
+      generatedByMetalsValues: Set[String]
+  ): (List[Enumerator], Option[Term], Set[String]) = {
+    val (perhapseValueNameAndNextQual, newMetalsNames) =
+      processValueNameAndNextQual(
+        termApply.args.head,
+        generatedByMetalsValues
+      )
+
+    termApply.fun match {
+      case termSelect: Term.Select
+          if termSelect.name.value == "flatMap" || termSelect.name.value == "map" =>
+        val shouldFlat = termSelect.name.value == "flatMap"
+
+        val result = for {
+          valueName <- perhapseValueNameAndNextQual.map(_._1)
+          nextQual <- perhapseValueNameAndNextQual.map(_._2)
+        } yield {
+          val (elems, maybeYieldTerm, updatedGenByMetalsVals) =
+            obtainNextYieldAndElemsForMap(
+              newMetalsNames,
+              perhapseLastName,
+              shouldFlat,
+              existingForElements,
+              currentYieldTerm,
+              termApply,
+              nextQual
+            )
+
+          processMap(
+            elems,
+            maybeYieldTerm,
+            updatedGenByMetalsVals,
+            valueName,
+            termSelect.qual
+          )
+        }
+        result.getOrElse(List.empty, None, generatedByMetalsValues)
+
+      case termSelect: Term.Select
+          if termSelect.name.value == "filter" || termSelect.name.value == "filterNot" ||
+            termSelect.name.value == "withFilter" =>
+        val isFilter =
+          termSelect.name.value == "filter" || termSelect.name.value == "withFilter"
+        processFilter(
+          perhapseValueNameAndNextQual,
+          newMetalsNames,
+          generatedByMetalsValues,
+          perhapseLastName,
+          isFilter,
+          existingForElements,
+          currentYieldTerm,
+          termSelect.qual
+        )
+
+      case _ =>
+        perhapseLastName match {
+          case Some(lastName) =>
+            (
+              Enumerator.Generator(
+                Pat.Var.apply(Term.Name.apply(lastName)),
+                termApply
+              )
+                +: existingForElements,
+              currentYieldTerm,
+              generatedByMetalsValues
+            )
+          case None =>
+            (existingForElements, Some(termApply), generatedByMetalsValues)
+        }
+    }
+  }
+
+  private def createNewName(
```

Comment:
Thank you so much for the code 💯 
I would be including that 👍 
I just disagree with the statement that 

>   I don't think it's needed to go all the way up to the top parent, but if users keep using same names this might happen, however this is a really edge scenario

Going up the parent is very low cost and adds lots of value. Let’s say we are processing two nested `ExtractValue` or `FlatMap` cases.

Reply:
Also, it helps in not having to always pass the whole Apply tree  of the whole chain of maps/flatMaps as an argument to some of the functions in the code, and instead just pass the most immediate tree we are processing at the moment. Let’s say a `Term.Select` we’re creating the name for.

Reply:
> Going up the parent is very low cost and adds lots of value. Let’s say we are processing two nested ExtractValue or FlatMap cases.

Going up the parent yes, but then traversing the entire tree is not low cost anymore. Especially if the tree is huge.

>Also, it helps in not having to always pass the whole Apply tree of the whole chain of maps/flatMaps as an argument to some of the functions in the code, and instead just pass the most immediate tree we are processing at the moment. Let’s say a Term.Select we’re creating the name for.

I am not sure how this pertains to the current topic. We need to create the MetalsNames object once and pass it around, but whichever tree in the chain we choose will have the same result.

Reply:
I mean the input size of the whole parent tree which gets traversed is going to be pretty small. Because a huge Scala file is going to have at most 2000 lines code.

Reply:
> I am not sure how this pertains to the current topic.
We need to add an extra argument to the methods in the code, in here.

Reply:
> I mean the input size of the whole parent tree which gets traversed is going to be pretty small. Because a huge Scala file is going to have at most 2000 lines code.

The tree is already large in that case, and I've seen bigger files.

> We need to add an extra argument to the methods in the code, in here.

What I am proposing is actually allowing us to remove some of it.

Reply:
Ok, I would implement and see.